File: scala/153.scala
object Solution {
    def findMin(nums: Array[Int]): Int = {
        if (nums.length == 1) nums.head 
        else if (nums.head < nums.last) nums.head
        else {
            val mid = nums.length / 2
            if (nums(mid) == nums.last) nums.last
            else if (nums(mid) < nums.head) findMin(nums.slice(0, mid+1))
            else findMin(nums.slice(mid, nums.length))
        }
    }
}

File: leetcodeJava/153.java
package leetcodeJava;

class Solution {
    public int findMin(int[] nums) {
        var n = nums.length;
        return find(nums, 0, n-1);
    }
    
    private int find(int[] nums, int l, int r) {
        if (l == r) return nums[l];
        else if (l == r-1) return Math.min(nums[l], nums[r]);
        if (nums[r] > nums[l]) return nums[l];
        var mid = (r-l)/2 + l;
        if (nums[mid] > nums[l]) return find(nums, mid+1, r);
        else return find(nums, l, mid);
    }
}
